src/librustc_save_analysis/lib.rs

```diff
@@ -1109,21 +1070,21 @@ fn id_from_def_id(id: DefId) -> rls_data::Id {
     rls_data::Id { krate: id.krate.as_u32(), index: id.index.as_u32() }
 }
 
-fn id_from_node_id(id: NodeId, scx: &SaveContext<'_, '_>) -> rls_data::Id {
-    let def_id = scx.tcx.hir().opt_local_def_id_from_node_id(id);
-    def_id.map(|id| id_from_def_id(id.to_def_id())).unwrap_or_else(|| {
-        // Create a *fake* `DefId` out of a `NodeId` by subtracting the `NodeId`
-        // out of the maximum u32 value. This will work unless you have *billions*
-        // of definitions in a single crate (very unlikely to actually happen).
-        rls_data::Id { krate: LOCAL_CRATE.as_u32(), index: !id.as_u32() }
-    })
+fn id_from_hir_id(id: hir::HirId, scx: &SaveContext<'_, '_>) -> rls_data::Id {
+    let def_id = scx.tcx.hir().opt_local_def_id(id);
+    def_id
+        .map(|id| id_from_def_id(id.to_def_id()))
```

Comment:
I'm having a hard time grokking the changes here. Why do we fall back to getting def_id from `id.owner`? IIUC the original  code worked around the assumption that node ids are inherently local and may only produce local defs.

Since we have `id: hir::HirId`, can't we directly go from `HirId` to (rls_data-like) `DefId`?

Reply:
I am not sure what exactly rls needs. Internally the compiler use `(opt_)local_def_id` to lookup the `LocalDefId` associated with some `HirId` because not all `LocalDefId` are `HirId` owners and especially because the real owner can be different from `hir_id.owner`. The fallback case (using `hir_id.owner`) should actually almost never happen in practice its more like a safe guard for very weird bugs in the compiler where a `HirId` does not actually have an associated `LocalDefId` at all.

Reply:
> I am not sure what exactly rls needs.

To be 100% frank I can't recall the reasons behind why we're creating `rls_data::Id`s like that :sweat_smile: 

> (...) its more like a safe guard for very weird bugs in the compiler where a `HirId` does not actually have an associated `LocalDefId` at all.

Could we document that somewhere here?

>  (...) because the real owner can be different from hir_id.owner (...)

Could you give an example, so I could better understand the situation?

Otherwise, this looks good! I'd personally very much appreciate appreciate a `match` with some specific comments like why there can be ever a `None` case or why `id.owner` is a good fallback :slightly_smiling_face: It's somewhat hard to follow along, otherwise

Reply:
Okay, so this sheds some more light as to why we use two-sided index. If we take a look at the save-analysis generated with this patch, we can see:
```
    {
      "attributes": [],
      "children": [],
      "decl_id": null,
      "docs": "",
      "id": {
        "index": 5,
        "krate": 0
      },
      "kind": "Local",
      "name": "world",
      "parent": null,
      "qualname": "world$18",
      "sig": null,
      "span": {
        "byte_end": 147,
        "byte_start": 142,
        "column_end": 14,
        "column_start": 9,
        "file_name": "src/tools/rls/tests/fixtures/common/src/main.rs",
        "line_end": 12,
        "line_start": 12
      },
      "value": "&str"
    },
    {
      "attributes": [],
      "children": [],
      "decl_id": null,
      "docs": "",
      "id": {
        "index": 5,
        "krate": 0
      },
      "kind": "Local",
      "name": "bar2",
      "parent": null,
      "qualname": "bar2$21",
      "sig": null,
      "span": {
        "byte_end": 207,
        "byte_start": 203,
        "column_end": 13,
        "column_start": 9,
        "file_name": "src/tools/rls/tests/fixtures/common/src/main.rs",
        "line_end": 15,
        "line_start": 15
      },
      "value": "Bar"
```

That's two "definitions", which have the same `def_id`. Informally, these are definitions in the name resolution sense (which don't have a separate DefId IIUC) and not item-like sense (which should have DefId). For the former, save-analysis used a negated index, so it created a *fake* def ids for those.

Because of this, I think we should retain the old behaviour for now (synthesizing `rls_data::Id using negative indexing of `HirId`/`NodeId` as a numeric value) in case we don't have a `LocalDefId` for a given `HirId`.

Reply:
Ok that makes perfect sense! I'll into that.
On a side note, i wondered if, before merging this, we should use all the fixture test files as test in rustc itself and generate the dumped analysis for all of those and checked for regressions. Iirc miri does a similar things. This way we would have a better confidence about the impact of the changes for this work.

Reply:
About `HirId` being owned by a different `LocalDefId` than `hir_id.owner`, you can take a look at my comment here: https://github.com/rust-lang/rust/issues/50928#issuecomment-635232363
TL;DR is closures behave like that

Reply:
That's a great idea in general. Because this is slowly becoming legacy and unloved code, there was an idea floating around that we should remove save_analysis at some point altogether (mb that's not the case anymore with this HIR rewrite?) - because of this, I'm not sure if this is worth the investment.

Addditionally, the IDs are not guaranteed to be stable (rn based on node ids as you've seen) so I'd be afraid this might induce more test update churn etc. If we can easily come up with a solution that verifies the relations and aux data defined in these files rather than concrete values, then I'd be up for it!

I wouldn't want to block this PR on it, though.

Reply:
I fixed the id generation and indeed it fixes `rls` tests

Reply:
Looks great, thanks!